legate/core/constraints.py

```diff
@@ -15,25 +15,46 @@
 from __future__ import annotations
 
 from collections.abc import Iterable
+from typing import TYPE_CHECKING, Any, Iterator, Optional, Sequence, Union
+
+from typing_extensions import Protocol
 
 from .partition import Restriction
 
+if TYPE_CHECKING:
+    from .store import Store
+
+
+class Expr(Protocol):
+    @property
+    def ndim(self) -> int:
+        ...
+
+    def subst(self, mapping: dict[PartSym, Expr]) -> Expr:
+        ...
+
+    def reduce(self) -> Expr:
+        ...
+
+    def unknowns(self) -> Iterator[Expr]:
+        ...
 
-class Expr:
-    def __eq__(self, rhs):
+    def __eq__(self, rhs: Expr) -> Alignment:  # type: ignore [override]
```

Comment:
The ignore is because `mypy` considers this an invalid override
```
legate37 ❯ mypy legate
legate/core/constraints.py: note: In member "__eq__" of class "Expr":
legate/core/constraints.py:42:5: error: Argument 1 of "__eq__" is incompatible with supertype "object"; supertype defines the argument type as "object"  [override]
        def __eq__(self, rhs: Expr) -> Constraint:
        ^
legate/core/constraints.py:42:5: note: This violates the Liskov substitution principle
legate/core/constraints.py:42:5: note: See https://mypy.readthedocs.io/en/stable/common_issues.html#incompatible-overrides
legate/core/constraints.py:42:5: note: It is recommended for "__eq__" to work with arbitrary objects, for example:
legate/core/constraints.py:42:5: note:     def __eq__(self, other: object) -> bool:
legate/core/constraints.py:42:5: note:         if not isinstance(other, Expr):
legate/core/constraints.py:42:5: note:             return NotImplemented
legate/core/constraints.py:42:5: note:         return <logic to compare two Expr instances>
legate/core/constraints.py:42:5: error: Return type "Constraint" of "__eq__" incompatible with return type "bool" in supertype "object"  [override]
```
Normally  this is almost always good advice, but given that this we are specifically essentially creating a DSL for deferred objects out of Python, this seems like one of the very few cases where it is sensible for `__eq__` to behave differently (so, reasonable to `ignore`)

We could, however, accept `object` but do a type check and raise an error to narrow the input type, if you think that is ok.